samples/nrf9160/http_application_update/src/main.c

```diff
@@ -0,0 +1,234 @@
+void bsd_irrecoverable_error_handler(uint32_t err)
+{
+	while (true) {
+
+	}
+}
+
+
+/**@brief Initialize application. */
+static void app_dfu_init(void)
+{
+	int i;
+
+	for (i = 0; i < FLASH_PAGE_MAX_CNT; i++) {
+		is_flash_page_erased[i] = false;
+	}
+	flash_dev = device_get_binding(DT_FLASH_DEV_NAME);
+	if (flash_dev == 0) {
+		printk("Nordic nRF flash driver was not found!\n");
+		return;
+	}
+
+	int retval = download_client_init(&dfu);
+
+	if (retval != 0) {
+		printk("download_client_init() failed, err %d", retval);
+		return;
+	}
+}
+
+
+/**@brief Start transfer of the file. */
+static void app_dfu_transfer_start(void)
+{
+	int retval = download_client_connect(&dfu);
+
+	if (retval != 0) {
+		printk("download_client_connect() failed, err %d",
+			retval);
+		return;
+	}
+
+	retval = download_client_start(&dfu);
+	if (retval != 0) {
+		printk("download_client_start() failed, err %d",
+			retval);
+		return;
+	}
+}
+
+
+static int app_download_client_event_handler(
+	struct download_client *const dfu,
+	enum download_client_evt event, u32_t error)
+{
+	int err;
+
+	switch (event) {
+	case DOWNLOAD_CLIENT_EVT_DOWNLOAD_FRAG: {
+		struct flash_pages_info info;
+		static u32_t address = CONFIG_FLASH_WRITE_OFFSET;
+
+		err = flash_get_page_info_by_offs(flash_dev, address, &info);
+		if (err != 0) {
+			printk("flash_get_page_info_by_offs returned error %d\n",
+				err);
+			return 1;
+		}
+		flash_write_protection_set(flash_dev, false);
+		if (!is_flash_page_erased[info.index]) {
+			err = flash_erase(flash_dev, address, info.size);
+			if (err != 0) {
+				printk("flash_erase returned error %d at address %08x\n",
+					err, address);
+				return 1;
+			}
+			is_flash_page_erased[info.index] = true;
+		}
+		err = flash_write(flash_dev, address, dfu->fragment,
+							dfu->fragment_size);
+		flash_write_protection_set(flash_dev, true);
+		if (err != 0) {
+			printk("Flash write error %d at address %08x\n",
+				err, address);
+			return 1;
+		}
+		address += dfu->fragment_size;
+		break;
+	}
+
+	case DOWNLOAD_CLIENT_EVT_DOWNLOAD_DONE:
+		printk("DOWNLOAD_CLIENT_EVT_DOWNLOAD_DONE");
+		break;
+
+	case DOWNLOAD_CLIENT_EVT_ERROR: {
+		download_client_disconnect(dfu);
+		printk("Download client error, please restart "
+				"the application\n");
+		break;
+	}
+	default:
+		break;
+	}
+	return 0;
+}
+
+
+/**@brief Turn on LED0 and LED1 if CONFIG_APPLICATION_VERSION
+ * is 2 and LED0 otherwise.
+ */
+static void led_app_version(void)
+{
+	struct device *dev;
+
+	dev = device_get_binding(LED_PORT);
+	if (dev == 0) {
+		printk("Nordic nRF GPIO driver was not found!\n");
+		return;
+	}
+
+	gpio_pin_configure(dev, LED0_GPIO_PIN, GPIO_DIR_OUT);
+	gpio_pin_write(dev, LED0_GPIO_PIN, 1);
+
+#if CONFIG_APPLICATION_VERSION == 2
+	gpio_pin_configure(dev, LED1_GPIO_PIN, GPIO_DIR_OUT);
+	gpio_pin_write(dev, LED1_GPIO_PIN, 1);
+#endif
+}
+
+void dfu_button_pressed(struct device *gpiob, struct gpio_callback *cb,
+		    u32_t pins)
+{
+	start_dfu = true;
```

Comment:
You may start a new transfer after some time. If a transfer is already in progress, it must first complete and the next one will be queued.

Reply:
Disabling button callback in this function now.